clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -672,6 +674,72 @@ check_function_markers()
         if (!run_checker(memrefs, false))
             return false;
     }
+    // Correctly skip return address check when the return address is
+    // unavailable.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1),
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_JUMP, TID, CALL_PC, CALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR, /*pc=*/123456),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly handle nested function calls.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1),
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID, CALL_PC, CALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR, CALL_PC + CALL_SZ),
+
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID, SECOND_CALL_PC,
+                           SECOND_CALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 3),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR,
+                       SECOND_CALL_PC + SECOND_CALL_SZ),
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID,
+                           /*pc=*/SECOND_CALL_PC + SECOND_CALL_SZ + 1, /*size=*/1),
+
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_JUMP, TID,
```

Comment:
Seems worth adding a comment about this being a tailcall.

Reply:
Done.

Reply:
Not seeing a comment for this line?

Reply:
I've updated the sequence a add comment to describe the jump  which jumps back to the beginning of the function, and have a function marker.